java-checks/src/main/java/org/sonar/java/checks/security/AndroidMobileDatabaseEncryptionKeysCheck.java

```diff
@@ -0,0 +1,121 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.security;
+
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ExpressionsHelper;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.checks.helpers.ExpressionsHelper.getSingleWriteUsage;
+
+@Rule(key = "S6301")
+public class AndroidMobileDatabaseEncryptionKeysCheck extends AbstractMethodDetection {
+
+  private static final String JAVA_LANG_STRING = "java.lang.String";
+
+  private static final MethodMatchers SQLITE_DATABASE_CONSTRUCTOR = MethodMatchers.create()
+    .ofTypes("net.sqlcipher.database.SQLiteDatabase")
+    .constructor()
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers SQLITE_DATABASE_METHODS = MethodMatchers.create()
+    .ofTypes("net.sqlcipher.database.SQLiteDatabase")
+    .names("changePassword", "openDatabase", "openOrCreateDatabase", "create")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY = MethodMatchers.create()
+    .ofTypes("io.realm.RealmConfiguration$Builder")
+    .names("encryptionKey")
+    .addParametersMatcher("byte[]")
+    .build();
+
+  private static final MethodMatchers STRING_TO_CHAR_GET_BYTES = MethodMatchers.create()
+    .ofTypes(JAVA_LANG_STRING)
+    .names("toCharArray", "getBytes")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(
+      SQLITE_DATABASE_CONSTRUCTOR,
+      SQLITE_DATABASE_METHODS,
+      REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY);
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    Arguments arguments = mit.arguments();
+    String messageArg = REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY.matches(mit) ? "encryptionKey" : "password";
```

Comment:
The downside of using `AbstractMethodDetection` is that you don't know which matcher matched. It means that you have to apply this matcher a second time. You can go for an `IssuableSubscriptionVisitor` and avoid redundant computation.